File: src/Stemmer.java

/*
   Porter stemmer in Java.
   Source : http://www.tartarus.org/~martin/PorterStemmer
*/


/**
 * Stemmer, implementing the Porter Stemming Algorithm
 *
 * The Stemmer class transforms a word into its root form.  The input
 * word can be provided a character at time (by calling add()), or at once
 * by calling one of the various stem(something) methods.
 */

class Stemmer {
    private char[] chars;
    private int beginningIndex, endingIndex, beginningIterator, endingIterator;
    public Stemmer() {
        chars = new char[50];
        beginningIndex = 0;
        endingIndex = 0;
    }

    /**
     * Add a character to the word being stemmed.  When you are finished
     * adding characters, you can call stem(void) to stem the word.
     */

    public void add(char character) {
        if (beginningIndex == chars.length) {
            char[] newChars = new char[beginningIndex + 50];
            System.arraycopy(chars, 0, newChars, 0, beginningIndex);
            chars = newChars;
        }
        chars[beginningIndex++] = character;
    }


    /**
     * After a word has been stemmed, it can be retrieved by toString(),
     * or a reference to the internal buffer can be retrieved by getResultBuffer
     * and getResultLength (which is generally more efficient.)
     */
    public String toString() { return new String(chars,0, endingIndex); }



    private boolean isConsonant(int index)
    {
        return switch (chars[index]) {
            case 'a', 'e', 'i', 'o', 'u' -> false;
            case 'y' -> index == 0 || !isConsonant(index - 1);
            default -> true;
        };
    }

    /** Consonant class measures the number of consonant sequences between 0 and beginningIterator. if c is
       a consonant sequence and v a vowel sequence, and <..> indicates arbitrary
       presence, for example:
          <c><v>       gives 0
          <c>vc<v>     gives 1
          <c>vcvc<v>   gives 2
          ....
    */
    class Consonant{
        int i = 0;
       public boolean checkConsonantForI(boolean doesCheckIsConsonant){
           while(true) {
               if (i > beginningIterator) return true;
               if (doesCheckIsConsonant == isConsonant(i)) break;
               i++;
           }
           i++;
           return false;
       }
       public int returnNumberOfConsonantSequences() {
           int n = 0;
           if (checkConsonantForI(false)) return n;
           while(true) {
               if (checkConsonantForI(true)) return n;
               n++;
               if (checkConsonantForI(false)) return n;
           }
       }
    }


    /** from0ToJAreVowels() is true <=> 0,...beginningIterator contains a vowel */

    private boolean from0ToJAreVowels() {
        for (int i = 0; i <= beginningIterator; i++){
            if (! isConsonant(i)) return true;
        }
        return false;
    }

    /** checkDoubleConsonantSequence(beginningIterator) is true <=> beginningIterator,(beginningIterator-1) contain a double consonant. */
    private boolean areTwoLettersEqual(int index){
        return chars[index] != chars[index-1];
    }
    private boolean checkDoubleConsonantSequence(int index) {
        if (index < 1 || areTwoLettersEqual(index)) return false;
        return isConsonant(index);
    }

   /** checkConsonantVowelConsonantSequence(i) is true <=> i-2,i-1,i has the form consonant - vowel - consonant
      and also if the second c is not w,x or y. this is used when trying to
      restore an e at the end of a short word. e.g.

         cav(e), lov(e), hop(e), crim(e), but
         snow, box, tray.

   */
    private boolean isSequenceWXY(int index){
        int character = chars[index];
        return character != 'w' && character != 'x' && character != 'y';
    }

    private boolean checkConsonantVowelConsonantSequence(int index) {
        if (index < 2 || !isConsonant(index) || isConsonant(index-1) || !isConsonant(index-2)) return false;
        return isSequenceWXY(index);

    }

    private boolean ends(String string) {
        int stringLength = string.length();
        int o = endingIterator - stringLength + 1;
        if (o < 0) return false;
        for (int i = 0; i < stringLength; i++)
            if (chars[o + i] != string.charAt(i))
                return false;
        beginningIterator = endingIterator - stringLength;
        return true;
    }

   /** adjustCharacterFromString(string) sets (beginningIterator + 1),...,endingIterator to
    * the characters in the string s, readjusting endingIterator. */

    private void adjustCharacterFromString(String string) {
        int l = string.length();
        for (int i = 0; i < l; i++)
            chars[i + beginningIterator + 1] = string.charAt(i);
        endingIterator = beginningIterator + l;
    }

    /** callAdjust(s) is used further down. */

    private void callAdjust(String s) {
        if (new Consonant().returnNumberOfConsonantSequences() > 0)
            adjustCharacterFromString(s);
    }

   /** changePlurals() gets rid of plurals and -ed or -ing. e.g.

          caresses  ->  caress
          ponies    ->  poni
          ties      ->  ti
          caress    ->  caress
          cats      ->  cat

          feed      ->  feed
          agreed    ->  agree
          disabled  ->  disable

          matting   ->  mat
          mating    ->  mate
          meeting   ->  meet
          milling   ->  mill
          messing   ->  mess

          meetings  ->  meet

   */
    private void changePluralsEndWithEdAndIng(){
        if ((ends("ed") || ends("ing")) && from0ToJAreVowels()) {
            endingIterator = beginningIterator;
            String[] strings = {"at", "bl", "iz"};
            for (String string : strings) {
                if (ends(string)){
                    adjustCharacterFromString(string + "e");
                    return;
                }
            }
            if (checkDoubleConsonantSequence(endingIterator)) {
                endingIterator--;
                int ch = chars[endingIterator];
                if (ch == 'l' || ch == 's' || ch == 'z') endingIterator++;
            }
            else if (new Consonant().returnNumberOfConsonantSequences() == 1
                    && checkConsonantVowelConsonantSequence(endingIterator)) adjustCharacterFromString("e");
        }
    }

    private void changePlurals() {
        if (chars[endingIterator] == 's') {
            if (ends("sses")) endingIterator -= 2;
            else if (ends("ies")) adjustCharacterFromString("i");
            else if (chars[endingIterator -1] != 's') endingIterator--;
        }
        if (ends("eed")) {
            if (new Consonant().returnNumberOfConsonantSequences() > 0) endingIterator--;
        }
        else changePluralsEndWithEdAndIng();
    }

    /** turnYToI() turns terminal y to i when there is another vowel in the stem. */

    private void turnYToI() {
        if (ends("y") && from0ToJAreVowels()) chars[endingIterator] = 'i';
    }


    private void doubleSuffixesFirstStep(char ch){
        switch (ch){
            case 'a'-> {if (ends("ational")) { callAdjust("ate"); break; }
                if (ends("tional")) callAdjust("tion");
            }
            case 'c'-> {if (ends("enci")) { callAdjust("ence"); break; }
                if (ends("anci")) callAdjust("ance");
            }
            case 'e' -> {if (ends("izer")) callAdjust("ize"); }

        }
    }
    private void doubleSuffixesSecondStep(char ch){
        switch (ch){
            case 'l' -> {if (ends("bli")) { callAdjust("ble"); break; }
                if (ends("alli")) { callAdjust("al"); break; }
                if (ends("entli")) { callAdjust("ent"); break; }
                if (ends("eli")) { callAdjust("e"); break; }
                if (ends("ousli")) callAdjust("ous");
            }
            case 'o' -> {if (ends("ization")) { callAdjust("ize"); break; }
                if (ends("ation")) { callAdjust("ate"); break; }
                if (ends("ator")) callAdjust("ate");
            }

        }
    }
    private void doubleSuffixesThirdStep(char ch){
        switch (ch) {
            case 's' -> {if (ends("alism")) { callAdjust("al"); break; }
                if (ends("iveness")) { callAdjust("ive"); break; }
                if (ends("fulness")) { callAdjust("ful"); break; }
                if (ends("ousness")) callAdjust("ous");
            }
            case 't' -> {if (ends("aliti")) { callAdjust("al"); break; }
                if (ends("iviti")) { callAdjust("ive"); break; }
                if (ends("biliti")) callAdjust("ble");
            }
            case 'g' -> {if (ends("logi")) callAdjust("log");}
        }
    }
    /** changeDoubleSuffix() maps double suffices to single ones. so -ization ( = -ize plus
     -ation) maps to -ize etc. note that the string before the suffix must give
     Consonant class output > 0. */
    private void changeDoubleSuffix() {
        if (endingIterator == 0) return;
        doubleSuffixesFirstStep(chars[endingIterator -1]);
        doubleSuffixesSecondStep(chars[endingIterator -1]);
        doubleSuffixesThirdStep(chars[endingIterator -1]);
    }

    /** changeSuffixFinish() deals with -ic-, -full, -ness etc. similar strategy to step3. */

    private void changeICAndFullSuffix() {
        switch (chars[endingIterator]) {
            case 'e' -> {if (ends("icate")) { callAdjust("ic"); break; }
                if (ends("ative")) { callAdjust(""); break; }
                if (ends("alize")) callAdjust("al");
            }
            case 'i' -> {if (ends("iciti")) callAdjust("ic"); }
            case 'l' -> {if (ends("ical")) { callAdjust("ic"); break; }
                if (ends("ful")) callAdjust("");
            }
            case 's' -> {if (ends("ness")) callAdjust(""); }
        }
    }



    private boolean changeSuffixFinishFirstStep(char ch){
        switch (ch) {
            case 'a' -> {
                return ends("al");
            }
            case 'c' -> {
                return ends("ance") || ends("ence");
            }
            case 'e' -> {
                return ends("er");
            }
            case 'i' -> {
                return ends("ic");
            }

        }
        return false;
    }
    private boolean changeSuffixFinishSecondStep(char ch){
        switch (ch){
            case 'l' -> {
                return ends("able") || ends("ible");
            }
            case 'n' -> {
                return ends("ant") || ends("ement") || ends("ment") || ends("ent");
            }
        }
        return false;
    }
    private boolean changeSuffixFinishThirdStep(char ch){
        switch (ch){
            case 'o' -> {
                return (ends("ion") && beginningIterator >= 0
                        && (chars[beginningIterator] == 's' || chars[beginningIterator] == 't')) || ends("ou");
            }
            /* takes care of -ous */
            case 's' -> {
                return ends("ism");
            }
            case 't' -> {
                return ends("ate") || ends("iti");
            }

        }
        return false;
    }
    private boolean changeSuffixFinishFourthStep(char ch){
        switch (ch) {
            case 'u' -> {
                return ends("ous");
            }
            case 'v' -> {
                return ends("ive");
            }
            case 'z' -> {
                return ends("ize");
            }
            default -> {
                return false;
            }
        }
    }
    /** changeSuffixFinish() takes off -ant, -ence etc., in context <c>vcvc<v>. */
    private void changeSuffixFinish() {
        if (endingIterator == 0) return;
        char ch = chars[endingIterator -1];
        boolean b = changeSuffixFinishFirstStep(ch) || changeSuffixFinishSecondStep(ch)
                || changeSuffixFinishThirdStep(ch) || changeSuffixFinishFourthStep(ch);
        if (b && (new Consonant().returnNumberOfConsonantSequences() > 1)) endingIterator = beginningIterator;
    }

    /** removeFinalE() removes a final -e if Consonant class output > 1. */
    private void removeFinalE() {
        beginningIterator = endingIterator;
        if (chars[endingIterator] == 'e') {
            int a = new Consonant().returnNumberOfConsonantSequences();
            if (a > 1 || a == 1 && !checkConsonantVowelConsonantSequence(endingIterator -1)) endingIterator--;
        }
        if (chars[endingIterator] == 'l' && checkDoubleConsonantSequence(endingIterator)
                && new Consonant().returnNumberOfConsonantSequences() > 1) endingIterator--;
    }

    /** Stem the word placed into the Stemmer buffer through calls to add().
     * Returns true if the stemming process resulted in a word different
     * from the input.  You can retrieve the result with
     * getResultLength()/getResultBuffer() or toString().
     */
    public void stem() {
        endingIterator = beginningIndex - 1;
        if (endingIterator > 1) {
            changePlurals();
            turnYToI();
            changeDoubleSuffix();
            changeICAndFullSuffix();
            changeSuffixFinish();
            removeFinalE();
        }
        endingIndex = endingIterator + 1;
        beginningIndex = 0;
    }


    /** stemWord gets a string and stems it to the output */
    public String stemWord(String string)
    {
        string = string.toLowerCase();
        char[] chars = string.toCharArray();
        for (char ch : chars) {
            if (Character.isLetter(ch)) {
                add(ch);
            }
        }
        stem();
        return toString();

    }
}

File: searchProject_csharp/SampleLibrary/SampleLibrary/Stemmer.cs
namespace SampleLibrary;

public class Stemmer
{
    private char[] wordArray;		// character array copy of the given string
	private int _stem, _end;			// indices to the current end (last letter) of the stem and the word in the array

	// Get the stem of a word at least three letters long:
	public string StemWord ( string word )
	{
		if ( string.IsNullOrWhiteSpace ( word ) || word.Length < 3 )
			return word;

		wordArray = word.ToCharArray ();
		_stem = 0;
		_end = word.Length - 1;

		CorrectPlurals ();
		ChangeLastYToI ();
		FoldDoubleSuffixes ();
		CorrectFullAndNess ();
		ChangeSuffixFinish ();
		RemoveFinalE ();

		return new string ( wordArray, 0, _end + 1 );
	}

	private void CorrectEndingWithEdAndIng()
	{
		_end = _stem;
		if ( EndsWith ( "at" ) )
			OverwriteEnding ( "ate" );
		else if ( EndsWith ( "bl" ) )
			OverwriteEnding ( "ble" );
		else if ( EndsWith ( "iz" ) )
			OverwriteEnding ( "ize" );
		else if ( EndsWithDoubleConsonant () )
		{
			if ( !"lsz".Contains ( wordArray [ _end - 1 ] ) )
				Truncate ();
		}
		else if ( ConsonantSequenceCount () == 1 && PrecededByCvc ( _end ) )
			OverwriteEnding ( "e" );
	}

	// Step 1: remove basic plurals and -ed/-ing:
	private void CorrectPlurals ()
	{
		if ( wordArray [ _end ] == 's' )
		{
			if ( EndsWith ( "sses" ) )
				Truncate ( 2 );
			else if ( EndsWith ( "ies" ) )
				OverwriteEnding ( "i" );
			else if ( wordArray [ _end - 1 ] != 's' )
				Truncate ();
		}

		if ( EndsWith ( "eed" ) )
		{
			if ( ConsonantSequenceCount () > 0 )
				Truncate ();
		}
		else if ( ( EndsWith ( "ed" ) || EndsWith ( "ing" ) ) && VowelInStem () )
		{
			CorrectEndingWithEdAndIng();
		}
	}

	// Step 2: change a terminal 'y' to 'i' if there is another vowel in the stem:
	private void ChangeLastYToI ()
	{
		if ( EndsWith ( "y" ) && VowelInStem () )
			OverwriteEnding ( "i" );
	}

	private void FoldDoubleSuffixesFirstStep(char ch)
	{
		switch (ch)
		{
			case 'a':
				if ( ReplaceEnding ( "ational", "ate" ) ) break;
				ReplaceEnding ( "tional", "tion" ); break;
			case 'c':
				if ( ReplaceEnding ( "enci", "ence" ) ) break;
				ReplaceEnding ( "anci", "ance" ); break;
			case 'e':
				ReplaceEnding ( "izer", "ize" ); break;
			case 'l':
				if ( ReplaceEnding ( "bli", "ble" ) ) break;
				if ( ReplaceEnding ( "alli", "al" ) ) break;
				if ( ReplaceEnding ( "entli", "ent" ) ) break;
				if ( ReplaceEnding ( "eli", "e" ) ) break;
				ReplaceEnding ( "ousli", "ous" ); break;
		}
	}

	private void FoldDoubleSuffixesSecondStep(char ch)
	{
		switch (ch)
		{
			case 'o':
				if ( ReplaceEnding ( "ization", "ize" ) ) break;
				if ( ReplaceEnding ( "ation", "ate" ) ) break;
				ReplaceEnding ( "ator", "ate" ); break;
			case 's':
				if ( ReplaceEnding ( "alism", "al" ) ) break;
				if ( ReplaceEnding ( "iveness", "ive" ) ) break;
				if ( ReplaceEnding ( "fulness", "ful" ) ) break;
				ReplaceEnding ( "ousness", "ous" ); break;
			case 't':
				if ( ReplaceEnding ( "aliti", "al" ) ) break;
				if ( ReplaceEnding ( "iviti", "ive" ) ) break;
				ReplaceEnding ( "biliti", "ble" ); break;
			case 'g':
				ReplaceEnding ( "logi", "log" ); break;
		}
	}

	// Step 3: fold double suffixes to single suffix, e.g., -ization = -ize + -ation -> -ize:
	private void FoldDoubleSuffixes ()
	{
		FoldDoubleSuffixesFirstStep(wordArray[_end - 1]);
		FoldDoubleSuffixesSecondStep(wordArray[_end - 1]);
	}

	// Step 4: replace -ic-, -full, -ness, etc. with simpler endings:
	private void CorrectFullAndNess ()
	{
		switch ( wordArray [ _end ] )
		{
			case 'e':
				if ( ReplaceEnding ( "icate", "ic" ) ) break;
				if ( ReplaceEnding ( "ative", "" ) ) break;
				ReplaceEnding ( "alize", "al" ); break;
			case 'i':
				ReplaceEnding ( "iciti", "ic" ); break;
			case 'l':
				if ( ReplaceEnding ( "ical", "ic" ) ) break;
				ReplaceEnding ( "ful", "" ); break;
			case 's':
				ReplaceEnding ( "ness", "" ); break;
		}
	}
	
	private bool changeSuffixFinishFirstStep(char ch){
        switch (ch) {
            case 'a' : {
                return EndsWith("al");
            }
            case 'c' : {
                return EndsWith("ance") || EndsWith("ence");
            }
            case 'e' : {
                return EndsWith("er");
            }
            case 'i' : {
                return EndsWith("ic");
            }

        }
        return false;
    }
    private bool changeSuffixFinishSecondStep(char ch){
        switch (ch){
            case 'l' : {
                return EndsWith("able") || EndsWith("ible");
            }
            case 'n' : {
                return EndsWith("ant") || EndsWith("ement") || EndsWith("ment") || EndsWith("ent");
            }
        }
        return false;
    }
    private bool changeSuffixFinishThirdStep(char ch){
        switch (ch){
            case 'o' : {
                return (EndsWith("ion") && _stem >= 0 && ( wordArray[_stem] == 's' || wordArray[_stem] == 't' )) || EndsWith("ou");
            }
            /* takes care of -ous */
            case 's' : {
                return EndsWith("ism");
            }
            case 't' : {
                return EndsWith("ate") || EndsWith("iti");
            }

        }
        return false;
    }
    private bool changeSuffixFinishFourthStep(char ch){
        switch (ch) {
            case 'u' : {
                return EndsWith("ous");
            }
            case 'v' : {
                return EndsWith("ive");
            }
            case 'z' : {
                return EndsWith("ize");
            }
            default : {
                return false;
            }
        }
    }
	// Step 5: remove -ant, -ence, etc.:
    private void ChangeSuffixFinish() {
        char ch = wordArray [ _end - 1 ];
        bool b = changeSuffixFinishFirstStep(ch) || changeSuffixFinishSecondStep(ch)
                || changeSuffixFinishThirdStep(ch) || changeSuffixFinishFourthStep(ch);
        if (b && ConsonantSequenceCount () > 1)
	        _end = _stem;
    }

    // Step 6: remove final 'e' if necessary:
	private void RemoveFinalE ()
	{
		_stem = _end;
		if ( wordArray [ _end ] == 'e' )
		{
			var m = ConsonantSequenceCount ();
			if ( m > 1 || m == 1 && !PrecededByCvc ( _end - 1 ) )
				Truncate ();
		}

		if ( wordArray [ _end ] == 'l' && EndsWithDoubleConsonant() && ConsonantSequenceCount () > 1 )
			Truncate ();
	}

	private void Truncate ( int n = 1 )
	{
		_end -= n;
	}
	
	// Count the number of CVC sequences:
	private int ConsonantSequenceCount ()
	{
		int m = 0, index = 0;
		for ( ; index <= _stem && IsConsonant ( index ); index++ ) ;
		if ( index > _stem )
			return 0;

		for ( index++; ; index++ )
		{
			for ( ; index <= _stem && !IsConsonant ( index ); index++ ) ;
			if ( index > _stem )
				return m;

			for ( index++, m++; index <= _stem && IsConsonant ( index ); index++ ) ;
			if ( index > _stem )
				return m;
		}
	}

	// Return true if there is a vowel in the current stem:
	private bool VowelInStem ()
	{
		for ( var i = 0; i <= _stem; i++ )
			if ( !IsConsonant ( i ) )
				return true;
		return false;
	}

	// Returns true if the character at the specified index is a consonant, with special handling for 'y':
	private bool IsConsonant ( int index )
	{
		if ( "aeiou".Contains ( wordArray [ index ] ) )
			return false;

		return wordArray [ index ] != 'y' || index == 0 || !IsConsonant ( index - 1 );
	}

	// Return true if the char. at the current index and the one preceeding it are the same consonant:
	private bool EndsWithDoubleConsonant()
	{
		return _end > 0 && wordArray [ _end ] == wordArray [ _end - 1 ] && IsConsonant ( _end );
	}

	// Check if the letters at i-2, i-1, i have the pattern: consonant-vowel-consonant (CVC) and the second consonant
	// is not w, x or y; used when restoring an 'e' at the end of a short word, e.g., cav(e), lov(e), hop(e), etc.:
	private bool PrecededByCvc ( int index )
	{
		if ( index < 2 || !IsConsonant ( index ) || IsConsonant ( index - 1 ) || !IsConsonant ( index - 2 ) )
			return false;

		return !"wxy".Contains ( wordArray [ index ] );
	}

	// Check if the given string appears at the end of the word:
	private bool EndsWith ( string s )
	{
		int length = s.Length, index = _end - length + 1;
		if ( index >= 0 )
		{
			for ( var i = 0; i < length; i++ )
				if ( wordArray [ index + i ] != s[i] )
					return false;

			_stem = _end - length;
			return true;
		}
		return false;
	}

	// Conditionally replace the end of the word:
	private bool ReplaceEnding ( string suffix, string s )
	{
		if ( EndsWith ( suffix ) && ConsonantSequenceCount () > 0 )
		{
			OverwriteEnding ( s );
			return true;
		}
		return false;
	}

	// Change the end of the word to a given string:
	private void OverwriteEnding ( string s )
	{
		int length = s.Length, index = _stem + 1;
		for ( var i = 0; i < length; i++ )
			wordArray [ index + i ] = s[i];
		_end = _stem + length;
	}
}